src/storage/src/source/healthcheck.rs

```diff
@@ -0,0 +1,809 @@
+                    SourceStatus::Starting,
+                    SourceStatus::Running,
+                    SourceStatus::Stalled,
+                    SourceStatus::Failed,
+                    SourceStatus::Dropped,
+                ],
+                true,
+            ),
+            (
+                SourceStatus::Starting,
+                vec![
+                    SourceStatus::Setup,
+                    SourceStatus::Running,
+                    SourceStatus::Stalled,
+                    SourceStatus::Failed,
+                    SourceStatus::Dropped,
+                ],
+                true,
+            ),
+            (
+                SourceStatus::Running,
+                vec![
+                    SourceStatus::Setup,
+                    SourceStatus::Starting,
+                    SourceStatus::Stalled,
+                    SourceStatus::Failed,
+                    SourceStatus::Dropped,
+                ],
+                true,
+            ),
+            (
+                SourceStatus::Stalled,
+                vec![
+                    SourceStatus::Setup,
+                    SourceStatus::Starting,
+                    SourceStatus::Running,
+                    SourceStatus::Failed,
+                    SourceStatus::Dropped,
+                ],
+                true,
+            ),
+            (SourceStatus::Failed, vec![SourceStatus::Dropped], true),
+            // Forbidden transitions
+            (SourceStatus::Setup, vec![SourceStatus::Setup], false),
+            (SourceStatus::Starting, vec![SourceStatus::Starting], false),
+            (SourceStatus::Running, vec![SourceStatus::Running], false),
+            (SourceStatus::Stalled, vec![SourceStatus::Stalled], false),
+            (
+                SourceStatus::Failed,
+                vec![
+                    SourceStatus::Setup,
+                    SourceStatus::Starting,
+                    SourceStatus::Running,
+                    SourceStatus::Stalled,
+                    SourceStatus::Failed,
+                ],
+                false,
+            ),
+            (
+                SourceStatus::Dropped,
+                vec![
+                    SourceStatus::Setup,
+                    SourceStatus::Starting,
+                    SourceStatus::Running,
+                    SourceStatus::Stalled,
+                    SourceStatus::Failed,
+                    SourceStatus::Dropped,
+                ],
+                false,
+            ),
+        ];
+
+        for test_case in test_cases {
+            run_test(test_case)
+        }
+
+        fn run_test(test_case: (SourceStatus, Vec<SourceStatus>, bool)) {
+            let (from_status, to_status, allowed) = test_case;
+            for status in to_status {
+                assert_eq!(allowed, from_status.can_transition(&status))
+            }
+        }
+    }
+
+    // Auxiliary functions
+    static PERSIST_CACHE: Lazy<Arc<Mutex<PersistClientCache>>> = Lazy::new(|| {
```

Comment:
@danhhz What do you think about these helpers for testing? Is that how we want other modules to test things that integrate with persist? Should we provide helpers with persist so that we don't have do duplicate these in the future?

Reply:
I copied most of this logic from the reclocking tests (with some snippets coming from other persist-related tests), so definite +1 from me on finding ways to not avoid all this copy-and-pasting (although my Rust-fu is not good enough to propose a way to make it better, as sharing test code between modules proved harder than I expected).

Reply:
My opinion is that unit tests should create one PeristClientCache/PersistClient at the top of the test and plumb it wherever needed. This lazy static thing was something petros wanted so he could share state (which did and still feels like a testing anti-pattern to me) 🤷